File: server/src/Models/DMToolkitModels/JoinTables/CharacterSpell.cs
using DMToolkit.API.Models.DMToolkitModels.Entities;
using DMToolkit.API.Models.DMToolkitModels.Instances;

namespace DMToolkit.API.Models.DMToolkitModels.JoinTables;

public class CharacterSpell : IJoinTable
{
    public required string CharacterId { get; set; }
    public Character Character { get; set; } = null!;
    public required string SpellId { get; set; }
    public Spell Spell { get; set; } = null!;
    public bool IsPrepared { get; set; } = false;
}

File: flutter-client/dm_toolkit/lib/models/dm_toolkit/join_tables/character_spell.dart
import 'package:dm_toolkit/models/dm_toolkit/entities/spell.dart';

class CharacterSpell {
  final Spell spell;
  final bool isPrepared;

  CharacterSpell({
    required this.spell,
    this.isPrepared = false
  });

  factory CharacterSpell.fromJson(Map<String, dynamic> json, List<Spell> spells) {
    try {
      var spellId = json['spellId'] as String;
      var spell = spells.firstWhere((s) => s.id == spellId);

      return CharacterSpell(
        spell: spell,
        isPrepared: json['isPrepared'] as bool
      );
    } catch (e) {
      throw FormatException('CharacterSpell model is invalid.', e);
    }
  }
}